backend/x11/backend.c

```diff
@@ -0,0 +1,374 @@
+		x11->screen->root_visual, mask, values);
+
+	output->surf = wlr_egl_create_surface(&x11->egl, &output->win);
+	if (!output->surf) {
+		wlr_log(L_ERROR, "Failed to create EGL surface");
+		return false;
+	}
+
+	init_atom(x11, &x11->atoms.wm_protocols, 1, "WM_PROTOCOLS");
+	init_atom(x11, &x11->atoms.wm_delete_window, 0, "WM_DELETE_WINDOW");
+
+	xcb_change_property(x11->xcb_conn, XCB_PROP_MODE_REPLACE, output->win,
+		x11->atoms.wm_protocols.reply->atom, XCB_ATOM_ATOM, 32, 1,
+		&x11->atoms.wm_delete_window.reply->atom);
+
+	xcb_map_window(x11->xcb_conn, output->win);
+	xcb_flush(x11->xcb_conn);
+
+	wl_signal_emit(&x11->backend.events.output_add, output);
+	wl_signal_emit(&x11->backend.events.input_add, &x11->keyboard_dev);
+	wl_signal_emit(&x11->backend.events.input_add, &x11->pointer_dev);
+
+	wl_event_source_timer_update(x11->frame_timer, 16);
+
+	return true;
+}
+
+static void wlr_x11_backend_destroy(struct wlr_backend *backend) {
+	if (!backend) {
+		return;
+	}
+
+	struct wlr_x11_backend *x11 = (struct wlr_x11_backend *)backend;
+
+	wl_event_source_remove(x11->frame_timer);
+	wlr_egl_free(&x11->egl);
+
+	xcb_disconnect(x11->xcb_conn);
+	free(x11);
+}
+
+static struct wlr_egl *wlr_x11_backend_get_egl(struct wlr_backend *backend) {
+	struct wlr_x11_backend *x11 = (struct wlr_x11_backend *)backend;
+	return &x11->egl;
+}
+
+bool wlr_backend_is_x11(struct wlr_backend *backend) {
+	return backend->impl == &backend_impl;
+}
+
+static struct wlr_backend_impl backend_impl = {
+	.start = wlr_x11_backend_start,
+	.destroy = wlr_x11_backend_destroy,
+	.get_egl = wlr_x11_backend_get_egl,
+};
+
+static void output_transform(struct wlr_output *output, enum wl_output_transform transform) {
+	// TODO
+}
+
+static void output_destroy(struct wlr_output *wlr_output) {
+	struct wlr_x11_output *output = (struct wlr_x11_output *)wlr_output;
+	struct wlr_x11_backend *x11 = output->x11;
+
+	eglDestroySurface(x11->egl.display, output->surf);
+	xcb_destroy_window(x11->xcb_conn, output->win);
+}
+
+static void output_make_current(struct wlr_output *wlr_output) {
+	struct wlr_x11_output *output = (struct wlr_x11_output *)wlr_output;
+	struct wlr_x11_backend *x11 = output->x11;
+
+	if (!eglMakeCurrent(x11->egl.display, output->surf, output->surf, x11->egl.context)) {
+		wlr_log(L_ERROR, "eglMakeCurrent failed: %s", egl_error());
+	}
+}
+
+static void output_swap_buffers(struct wlr_output *wlr_output) {
+	struct wlr_x11_output *output = (struct wlr_x11_output *)wlr_output;
+	struct wlr_x11_backend *x11 = output->x11;
+
+	if (!eglSwapBuffers(x11->egl.display, output->surf)) {
+		wlr_log(L_ERROR, "eglSwapBuffers failed: %s", egl_error());
+	}
+}
+
+static struct wlr_output_impl output_impl = {
+	.transform = output_transform,
+	.destroy = output_destroy,
+	.make_current = output_make_current,
+	.swap_buffers = output_swap_buffers,
+};
+
+static void input_destroy(struct wlr_input_device *input_dev) {
+	// Do nothing
+}
```

Comment:
You don't have to provide implementations if they don't actually do anything. Each type should be checking if the impl exists and the function exists before calling it.